app/uk/gov/hmrc/tai/service/IncomeService.scala

```diff
@@ -87,7 +87,7 @@ class IncomeService @Inject()(
     incomes(nino, TaxYear()).map(_.nonTaxCodeIncomes.untaxedInterest)
 
   def taxCodeIncomes(nino: Nino, year: TaxYear)(implicit hc: HeaderCarrier): Future[Seq[TaxCodeIncome]] =
-    incomeRepository.taxCodeIncomes(nino, year)
+    incomeRepository.taxCodeIncomes(nino, year).map(_.map(t => t.copy(taxCode = t.taxCodeWithEmergencySuffix)))
```

Comment:
I don't really like this way of doing this. My main concerns is that to get a _correct_ model we have to go through this method and if we were to call `taxCodeWithEmergencySuffix` after retrieving the model we would get incorrect data as it would could potentially have the emergency suffix appended twice. 

A better solution in my opinion would be have some private field on the `TaxCodeIncome` object that represent the  base tax code and have `taxCode` as a lazy val that returns the base tax code appened with the suffix if applicable.

```
case class TaxCodeIncome(
  componentType: TaxComponentType,
  ...
  private baseTaxCode: String,
  ...
  basisOperation: BasisOperation,
  ...) {

   lazy val taxCode: String = basisOperation match {
    case Week1Month1BasisOperation => baseTaxCode + TaiConstants.EmergencyTaxCode
    case _                         => baseTaxCode
  }
}
```

Reply:
Hi Kurt,

A couple issues we are thinking of (unless we've misunderstood).. without changing the JSON writes for this model this won't give us what is needed (although it would be fine for using the model within the service).

With this, the JSON returned from the backend will be `baseTaxCode` and also the emergency tax code logic will not have been applied to it.. hence we applied it in the `taxCodeIncomes` method.

Unless we change the writers on the backend? Although there seems to be some inconsistency on the backend with the writers for the `TaxCodeIncome` model.